src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs

```diff
@@ -2,67 +2,95 @@
+    public virtual bool ShouldInvoke(HttpContext context)
     {
-        ArgumentGuard.NotNull(endpoint);
-        ArgumentGuard.NotNull(logger);
-        Endpoint = endpoint;
+        ArgumentGuard.NotNull(context);
+        ManagementEndpointOptions mgmtOptions = ManagementEndpointOptionsMonitor.GetFromContextPath(context.Request.Path, out string managementContextName);
+        HttpMiddlewareOptions endpointOptions = EndpointHandler.Options;
+        bool enabled = endpointOptions.IsEnabled(mgmtOptions);
+        bool exposed = endpointOptions.IsExposed(mgmtOptions);
+
+        bool isCFContext = managementContextName == CFContext.Name;
+        _logger.LogDebug($"endpointHandler: {endpointOptions.Id}, contextPath: {context.Request.Path}, enabled: {enabled}, exposed: {exposed}");
+        return enabled && (exposed || isCFContext);
     }
 
-    public virtual string HandleRequest()
+    public async Task InvokeAsync(HttpContext context, RequestDelegate next)
     {
-        TResult result = Endpoint.Invoke();
-        return Serialize(result);
+        ArgumentGuard.NotNull(context);
+
+        if (ShouldInvoke(context))
+        {
+            context.HandleContentNegotiation(_logger);
+            TResult result = await InvokeEndpointHandlerAsync(context, context.RequestAborted);
+            await WriteResponseAsync(result, context, context.RequestAborted);
+        }
+        else
+        {
+            // Terminal middleware
+            context.Response.StatusCode = (int)HttpStatusCode.NotFound;
+        }
     }
 
-    public virtual string Serialize(TResult result)
+    protected abstract Task<TResult> InvokeEndpointHandlerAsync(HttpContext context, CancellationToken cancellationToken);
+
+    protected virtual async Task WriteResponseAsync(TResult result, HttpContext context, CancellationToken cancellationToken)
     {
-        try
-        {
-            JsonSerializerOptions serializerOptions = managementOptions.CurrentValue.SerializerOptions;
-            JsonSerializerOptions options = GetSerializerOptions(serializerOptions);
+        ArgumentGuard.NotNull(context);
 
-            return JsonSerializer.Serialize(result, options);
-        }
-        catch (Exception e) when (e is ArgumentException or ArgumentNullException or NotSupportedException)
+        if (Equals(result, null))
         {
-            logger.LogError(e, "Error serializing {MiddlewareResponse}", result);
+            return;
         }
 
-        return string.Empty;
+        JsonSerializerOptions options = GetSerializerOptions();
+        await JsonSerializer.SerializeAsync(context.Response.Body, result, options, cancellationToken);
     }
 
-    internal static JsonSerializerOptions GetSerializerOptions(JsonSerializerOptions serializerOptions)
+    protected virtual JsonSerializerOptions GetSerializerOptions()
     {
+        JsonSerializerOptions serializerOptions = ManagementEndpointOptionsMonitor.CurrentValue.SerializerOptions;
+
```

Comment:
Please remove the addition of endpoint-specific converters below (`HealthConverterV3`, `MetricsResponseConverter`) from this class and place them in overrides of `GetSerializerOptions()` in the specific endpoint middleware classes. (follow-up on [this conversation](https://github.com/SteeltoeOSS/Steeltoe/pull/1130#discussion_r1226398669))

Reply:
This is not possible because this needs to happen before any serialization was done. I tried this earlier.

Reply:
I don't understand why that does not work. Moving the logic into an overridden method does not change the order of execution, right? The overridden method would call `base()`, then add its extra converter. Any callers of other virtual `GetSerializerOptions() ` would observe the same options as before.

Reply:
This fails because the `JsonSerializerOptions` instance gets frozen after the first usage (handling a request), so adding a converter later on throws. This could be worked around by making a copy on each request, but that's expensive, so let's leave it as-is.